Hook for managing directories and files in Azure Data Lake Storage Gen2

bharanidharan14 opened the pull request:
Created hook for supporting ADLS gen2, which uses the WASB connection and connects to ADLS gen2 storage

Relates to https://github.com/apache/airflow/issues/28223

bharanidharan14 committed:
ADLS gen2 hook

ADLS gen2 hook

bharanidharan14 committed:
ADLS gen2 hook

ADLS gen2 hook

tatiana commented on airflow/providers/microsoft/azure/hooks/adls.py:
```diff
@@ -0,0 +1,221 @@
+from airflow.hooks.base import BaseHook
+
+
+class AzureDataLakeStorageV2(BaseHook):
```
```suggestion
class AzureDataLakeStorageV2Hook(BaseHook):
```

bharanidharan14 replied:
Now I changed the Hook name now.

eladkal replied:
I'm not sure about this.
eventually v2 will be the standard and on that day it would be redundant to mention v2.

I see that all current v1 services are listed as `AzueDataLakeStorage*`
WDYT about having the new ones as `Adls*`?

bharanidharan14 replied:
Yes, the current  v1 services are listed as AzueDataLakeStorage. Now it is `AzureDataLakeStorageClient` can it be `AzureDataLakeStorageClientHook`. WDYT?

eladkal replied:
since we don't have hook for adls it can be anything we want.
i suggested the pattern: `Adls*` so:


| Current V1         | Suggested V2       |
|--------------------|--------------------|
|          -         | AdlsClientHook     |
| ADLSDeleteOperator | AdlsDeleteOperator |
| ADLSListOperator   | AdlsListOperator   |

eladkal replied:
Alternatively we can also do:
```
microsoft/azure/hooks/adsl_v2.py
microsoft/azure/operators/adsl_v2.py
```
then class names can have the same names as v1 because they are not in the same file
and once v1 is removed we can change the files names which is easier change then changing classes names.

tatiana commented on airflow/providers/microsoft/azure/hooks/adls.py:
```diff
@@ -0,0 +1,221 @@
+            return DataLakeServiceClient(
+                account_url=f"https://{conn.login}.dfs.core.windows.net", credential=token_credential, **extra
+            )
+        credential = conn.password
```
Would it make sense to do something similar to:
https://github.com/apache/airflow/blob/main/airflow/providers/microsoft/azure/hooks/wasb.py#L202?

I wonder how much in common the Wasb & the AzureDataLakeStorageV2 implementation have - and if it would make sense to have a `BaseAzureHook`

bharanidharan14 replied:
@tatiana I am using WASB connection details to connect to ADLS gen2 storage account, apart the connection details nothing is similar I think

tatiana replied:
That's a shame, @bharanidharan14 - I didn't realise things were so disconnected in Azure

bharanidharan14 replied:
@tatiana ADLS gen2 is using different protocols and APIs and different client-type principles to connect. Currently, I have created separate connection types and connection details.

bharanidharan14 committed:
Add Test case

bharanidharan14 committed:
Add Test case

bharanidharan14 committed:
Moved hook to existing data lake hook file

Fix mypy issues

Fix mypy issues

Fix mypy issues

Fix mypy issues

Add adls v2 connection

Fix doc

Fix doc

Moved hook to existing data lake hook file

Moved hook test case

Moved hook test case

Fix doc

bharanidharan14 committed:
Moved hook to existing data lake hook file

Fix mypy issues

Fix mypy issues

Fix mypy issues

Fix mypy issues

Add adls v2 connection

Fix doc

Fix doc

Moved hook to existing data lake hook file

Moved hook test case

Moved hook test case

Fix doc

eladkal commented on airflow/providers/microsoft/azure/hooks/adls.py:
```diff
@@ -0,0 +1,264 @@
+from airflow.providers.microsoft.azure.hooks.wasb import WasbHook
+
+
+class AzureDataLakeStorageClient(WasbHook):
```
https://github.com/apache/airflow/issues/28223 description states that wasb is legacy if so what is the motivation to inhert from wsab for adsl gen2?

bharanidharan14 replied:
Only for connection details, I thought of using these WASB connection details. should I create a new connection ?

eladkal replied:
If wasb is legacy and we expect it to be removed some day then I don't think we utlize it for other services
but I'm not Azure expert - I leave this desicion to people who knows the service.

What i do want to have here is docs and specifically connection docs that explains how to make this work.
(I want to avoid questions raised down the road by users who will ask to refactor because we utlize legacy service)

bharanidharan14 replied:
Ok let me add new connection, with docs

tatiana replied:
@bharanidharan14 @eladkal  did we consider having an `AzureBaseHook` similar to `GoogleBaseHook`
https://github.com/apache/airflow/blob/main/airflow/providers/google/common/hooks/base_google.py
This way, both `AzureDataLakeStorageHook` and `WasbHook` could inherit from it

luanmorenomaciel replied:
@tatiana and @mateusoliveiraowshq as per our conversation, implementing the `AzureBaseHook` would be challenging since Azure unfortunately does not share the same connection client principles (BlobServiceClient & DataLakeServiceClient) between the WASB and ABFS protocol.

[Azure Blob Storage - WASB ](https://pypi.org/project/azure-storage-blob/)
[Azure Data Lake Storage Gen2 - ABFS](https://pypi.org/project/azure-storage-file-datalake/)

What we could do instead of is to have the following hook structure:

[airflow.providers.microsoft.azure.hooks.wasb](https://airflow.apache.org/docs/apache-airflow-providers-microsoft-azure/stable/_api/airflow/providers/microsoft/azure/hooks/wasb/index.html) (Already Exists)

[airflow.providers.microsoft.azure.hooks.azure_data_lake](https://airflow.apache.org/docs/apache-airflow-providers-microsoft-azure/1.0.0/_api/airflow/providers/microsoft/azure/hooks/azure_data_lake/index.html#module-airflow.providers.microsoft.azure.hooks.azure_data_lake) (Suggested)

Meaning that as per @bharanidharan14 @eladkal conversation we would recommend to inherid the hook from the **airflow.providers.microsoft.azure.hooks.azure_data_lake** instead of **airflow.providers.microsoft.azure.hooks.wasb**, that would make more sense since WASB protocol has been marked as legacy hence the, implementation process are different from each other.



![wasb-abfs](https://user-images.githubusercontent.com/20648427/208137163-382b616f-d14b-413c-afa6-0c7506db95e6.png)

bharanidharan14 replied:
@luanmorenomaciel I don't think inheriting from the **airflow.providers.microsoft.azure.hooks.azure_data_lake** is also not recommended because this ADLS existing hook is gen1 and it's being getting retired from feb. They may stop supporting for that too as well. so its better to inherit from `BaseHook`. WDYT ?

luanmorenomaciel replied:
We should be fine to inherit from BaseHook

bharanidharan14 replied:
@luanmorenomaciel Currently I am inheriting from  BaseHook. Can you also review this PR

tatiana commented on airflow/providers/microsoft/azure/hooks/adls_v2.py:
```diff
@@ -0,0 +1,307 @@
+from hooks.base import BaseHook
+
+
+class AdlsClientHook(BaseHook):
```
Would it make sense for this to be defined inside `airflow/providers/microsoft/azure/hooks/azure_data_lake`?

kaxil replied:
@bharanidharan14 ^^

bharanidharan14 replied:
Addressed it

tatiana commented on airflow/providers/microsoft/azure/hooks/adls_v2.py:
```diff
@@ -0,0 +1,307 @@
+    .. seealso::
+        https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-directory-file-acl-python
+
+    :param adls_v2_conn_id: Reference to the :ref:`adls_v2 connection <howto/connection:adls_v2>`.
```
Hi @bharanidharan14 , it would be great if we could rename things as follows:

* `airflow/providers/microsoft/azure/hooks/adls_v2.py` -> `airflow/providers/microsoft/azure/hooks/azure_data_lake.py` (use existing module)
* `AdlsClientHook` -> AzureDataLakeStorageV2Hook` (be more explicit since we'll have V1 and V2 ADLS hooks coexisting for some time in Airflow/Azure, it is probably worth having V2 explicit in the class name)
* `adls_v2_conn_id` -> `adls_conn_id` (if the class name has V2, we can document that this parameter should be a V2 connection ID, but I think the parameter could be called only adls_conn_id)

bharanidharan14 replied:
@tatiana Will address this and let you know

bharanidharan14 replied:
@tatiana Addressed your suggestion

bharanidharan14 replied:
@tatiana Can you review it now

tatiana reviewed (commented):
@bharanidharan14 we're very close to merging this work.

There is one comment pending: https://github.com/apache/airflow/pull/28262/files#r1056265305
Rebasing and passing tests also seem to be pending.

eladkal commented on tests/providers/microsoft/azure/hooks/test_adls_v2.py:
```diff
@@ -0,0 +1,108 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
If the hook logic is in `hooks/data_lake.py` why do we create a `test_adls_v2.py` ?

bharanidharan14 replied:
Yeah @eladkal I am moving the tests to `test_azure_data_lake` file. working on it

bharanidharan14 replied:
Moved the test to `test_azure_data_lake` file

bharanidharan14 replied:
@eladkal Can you review it now

bharanidharan14 committed:
Merge branch 'adls-gen2' of https://github.com/astronomer/airflow into adls-gen2
